desc/magnetic_fields.py

```diff
@@ -795,3 +831,464 @@ def odefun(rpz, s):
+        object's R_lmn, Z_lmn etc instead of any passed-in values
+
+    """
+
+    _io_attrs_ = (
+        MagneticField._io_attrs_
+        + FourierRZToroidalSurface._io_attrs_
+        + [
+            "_surface_grid",
+            "_params",
+        ]
+    )
+
+    def __init__(
+        self,
+        potential,
+        surface_grid,
+        params={},
+        potential_dtheta=None,
+        potential_dzeta=None,
+        R_lmn=None,
+        Z_lmn=None,
+        modes_R=None,
+        modes_Z=None,
+        NFP=1,
+        sym="auto",
+        rho=1,
+        name="",
+        check_orientation=True,
+        surface=None,
+    ):
+        self._potential = potential
+        self._surface_grid = surface_grid
+        self._params = params
+
+        if surface:
+            if not isinstance(surface, FourierRZToroidalSurface):
+                raise TypeError(
+                    "Expected type FourierRZToroidalSurface for argument surface, "
+                    f"instead got type {type(surface)}"
+                )
+            R_lmn = surface.R_lmn
+            Z_lmn = surface.Z_lmn
+            modes_R = surface._R_basis.modes[:, 1:]
+            modes_Z = surface._Z_basis.modes[:, 1:]
+            NFP = surface.NFP
+            sym = surface.sym
+            rho = surface.rho
+
+        assert surface_grid.NFP == NFP, "NFP of surface must match NFP of surface_grid"
+
+        super().__init__(
+            R_lmn, Z_lmn, modes_R, modes_Z, NFP, sym, rho, name, check_orientation
+        )
+        self._set_derivatives(potential_dtheta, potential_dzeta)
+
+    @property
+    def surface_grid(self):
+        """Grid of points to evaluate surface current at."""
+        return self._surface_grid
+
+    @surface_grid.setter
+    def surface_grid(self, new):
+        if new != self._surface_grid:
+            self._surface_grid = new
+
+    @property
+    def params(self):
+        """Dict of parameters to pass to potential function and its derivatives."""
+        return self._params
+
+    @params.setter
+    def params(self, new):
+        if new != self._params:
+            if len(new) != len(self._params):
+                raise UserWarning(
+                    "Length of new params is different from length of current params! "
+                    "May cause errors unless potential function is also changed."
+                )
+            self._params = new
+
+    @property
+    def potential(self):
+        """Potential function, signature (theta,zeta,**params) -> potential value."""
+        return self._potential
+
+    @potential.setter
+    def potential(self, new):
+        if new != self._potential:
+            self._potential = new
+            # reset derivatives if potential has changed, using AD
+            self._set_derivatives()
+
+    def _set_derivatives(self, potential_dtheta=None, potential_dzeta=None):
+        if potential_dtheta:
+            self._potential_t = partial(potential_dtheta, **self.params)
```

Comment:
if you close over `params` here we won't be able to differentiate wrt them

Reply:
True but I cannot vmap it in the next line if we dont close over the params. Maybe not even worth keeping the AD one. I forgot why we need to use vmap for the auto derivative though

Reply:
removed the AD part as it was complicated to deal with when using the data index